libraries/AP_RangeFinder/AP_RangeFinder_DFRobot_Lidar07.cpp

```diff
@@ -0,0 +1,241 @@
+// measurement message: temperature
+#define LIDAR07_TEMP_OFFSET             (LIDAR07_DIST_OFFSET + sizeof(uint16_t))
+// measurement message: amplitude
+#define LIDAR07_AMP_OFFSET              (LIDAR07_TEMP_OFFSET + sizeof(uint16_t))
+// measurement message: ambient light
+#define LIDAR07_AMB_OFFSET              (LIDAR07_AMP_OFFSET + sizeof(uint16_t))
+// measurement message: ToF phase
+#define LIDAR07_TOF_OFFSET              (LIDAR07_AMB_OFFSET + sizeof(uint16_t))
+// crc32 for controller to sensor
+#define LIDAR07_CRC_OFFSET              (LIDAR07_VALUE_OFFSET + sizeof(uint32_t))
+// crc32 for sensor to given payload size
+#define LIDAR07_CRC_SENSOR_OFFSET(n)    (LIDAR07_DATA_OFFSET + n)
+// Total length for controller to sensor
+#define LIDAR07_CONTROLLER_LEN          (LIDAR07_CRC_OFFSET + sizeof(uint32_t))
+// Total length for sensor to controller
+#define LIDAR07_SENSOR_LEN(n)           (LIDAR07_CRC_SENSOR_OFFSET(n) + sizeof(uint32_t))
+// Longest payload length
+#define LIDAR07_MAX_LEN                 (16)
+
+// the sensor supports rates up to 100 Hz
+// use the lowest common rate between Copter and Plane: 20 Hz = 50 ms
+#define LIDAR07_FRAME_INTERVAL_MS       (50)
+
+// Invalidate the temperature and other readings after 2 measurement intervals
+#define LIDAR07_DATA_EXPIRATION_MS      (2*LIDAR07_FRAME_INTERVAL_MS)
+
+// Retry initialization if inactive for 30 seconds
+#define LIDAR07_STALE_CONNECTION_MS     (30*1000)
+
+// DFRobot's CRC32 implementation uses unreflected bit order
+// apparently it's not the way CRC32 is used elsewhere so it makes sense to hide
+// this utility under the define block
+template <typename T> T reflect(const T& p)
+{
+    // NOTE: straightforward bit reflection, can probably be optimized if used in high frequency code
+    T res = 0;
+    size_t bit_len = sizeof(T) * 8;
+    for (size_t i = 0; i < bit_len; i++) {
+        res |= bool(p & (1<<i)) << (bit_len - i - 1);
+    }
+    return res;
+}
+
+uint32_t AP_RangeFinder_DFRobot_Lidar07::crc32_unreflected(uint8_t *buf, size_t n)
+{
+    uint8_t ubuf[n];
+    for (size_t i = 0; i < n; i++)
+        ubuf[i] = reflect<uint8_t>(buf[i]);
+    return reflect<uint32_t>(crc_crc32(0xffffffff, ubuf, n));
+}
+
+void AP_RangeFinder_DFRobot_Lidar07::send_command(uint8_t reg, uint32_t value)
+{
+    if (uart == nullptr) {
+        return;
+    }
+
+    uint8_t cmd_buf[LIDAR07_CONTROLLER_LEN];
+    cmd_buf[LIDAR07_SOF_OFFSET] = LIDAR07_SOF_CONTROLLER;
+    cmd_buf[LIDAR07_REG_OFFSET] = reg;
+    *reinterpret_cast<uint32_t*>(&cmd_buf[LIDAR07_VALUE_OFFSET]) = htole32(value);
+    *reinterpret_cast<uint32_t*>(&cmd_buf[LIDAR07_CRC_OFFSET]) = htole32(crc32_unreflected(cmd_buf, LIDAR07_CRC_OFFSET));
+    uart->write(cmd_buf, sizeof(cmd_buf));
+
+    // the sensor doesn't quite like when we lump commands together, give it piecewise input
+    uart->flush();
+
+    // as well as yield to the scheduler to give the sensor some pause to think
+    hal.scheduler->delay(50);
```

Comment:
Can't sleep here - this is the main thread!

Reply:
Hmm is it? I mean I specifically avoided doing that in `init_serial` for that reason. 
I mean here it's scheduled as a separate task: https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/ArduPlane.cpp#L85

So what I assumed would happen is when I do `delay()` the rangefinder thread yields before the next rescheduling. Am I wrong in this assumption?

This is also happening with MAVLink FTP hence why I thought it'd be acceptable: https://github.com/ArduPilot/ardupilot/blob/master/libraries/GCS_MAVLink/GCS_FTP.cpp#L152

Reply:
Deleted my previous message after having re-read how scheduling work. I'll make a state machine here then.